Supporting comma separated values in environment variables on command-line

arjunrao87 opened the pull request:
Based on the conversation in #308 , this PR adds the ability to pass multiple comma separated env variables for a single env flag. This is done by replacing StringSlice with StringArray.

Please let me know if I need to do something else to get this PR merged into master.

Thanks!

andresmgot commented:
I tested the changes, it works for me:
 - Using StringSlice:
```
$ kubeless function deploy hello --env foo=bar,hap ...
$ kubectl exec -it hello-1730543884-tw4nj env | grep foo
foo=bar
$ kubectl exec -it hello-1730543884-tw4nj env | grep hap
hap=
```
 - Using StringArray:
```
$ kubeless function deploy hello --env foo=bar,hap ...
$ kubectl exec -it hello-3941471439-m0ht6 env | grep foo
foo=bar,hap
```
Maybe we should do the same for the `labels` option, what do you think @sebgoa?

arjunrao87 commented:
@andresmgot I thought the same thing but I wasn't sure what the expected interaction with labels was. If you guys agree its needed, I can add that change to this PR as well.

andresmgot commented:
As far as I can see there are no reasons to use `StringSlice` since it make impossible to use commas in labels or env vars. Could you update the PR with that?

arjunrao87 commented:
Added labels change to the PR as well

andresmgot commented:
Sorry, checking the Travis result, there is a test that enforces the opposite:
```
# time="2017-09-27T14:07:23Z" level=fatal msg="Function.k8s.io \"get-python-metadata\" is invalid: metadata.labels: Invalid value: \"bar,bar=foo,foobar\": a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')" 
```
So, to avoid breaking changes please revert c0b3ec3. I will merge this PR afterwards.

Thanks for checking!

arjunrao87 commented:
The tests pass now. @andresmgot  please let me know if theres anything else I need to do to get the PR merged. Thanks!

ngtuna commented:
Thanks @arjunrao87 for the PR. Can you also add a unit test for this ?

arjunrao87 commented:
@ngtuna Added tests for deploy.go and update.go. Please let me know if any changes are required.

anguslees commented:
@arjunrao87 so we want to test that commas in environment values work, so this doesn't break again in the future.  We should a) add a case to `function_test.go` to make sure `parseEnv` does the right thing (easy, just put back the thing you had earlier), but the other important half is b) whether the command line parsing does the right thing.

You can test the command line parsing directly with (eg)
```golang
updateCmd.SetArgs([]string{"update", "foo", "--env", "foo:bar,hap", "--env", "x:y"})
```
then check for `updateCmd.Flags().GetStringArrayP("env")` with the expected value(s).  Similar for deploy.

This feels a bit too implementation specific and fragile to me, and better (imo) would be an integration test that actually invoked kubeless cli and looked for the expected environment values - but that's a _much_ larger change so I'm not suggesting doing that as part of this PR unless you're feeling particularly eager.

anguslees commented:
this lgtm, is there anything still "WIP" about this?

arjunrao87 commented:
There shouldnt be anything WIP anymore. Updated the issue title. I have updated the tests based on what we discussed. Please let me know if there is something else that needs to be done to get the PR merged. I wont be able to to add the integration test at the moment but I can open an issue for it to take a look at it in the near future if that works.
